### ISSUE TITLE
"fieldPath: status.podIP" may use wrong IP family in dual stack cluster.

### ISSUE THREAD

**Issue by Jean-Daniel** [kubernetes/kubernetes#103263]
#### What happened:

In a dual-stack cluster with IPv4 as default family (first kubelet IP), a container is created with a env var set to  `fieldPath: status.podIP`.

When the container is running, `status.podIP` properly contains the IPv4

```yaml
apiVersion: v1
kind: Pod
metadata:
  …
spec:
  containers:
  - args:
    - --config.file=/etc/alertmanager/config/alertmanager.yaml
    - --storage.path=/alertmanager
    - --data.retention=120h
    - --cluster.listen-address=[$(POD_IP)]:9094
    - --web.listen-address=:9093
    - --web.route-prefix=/
    - --cluster.peer=alertmanager-main-0.alertmanager-operated:9094
    - --cluster.peer=alertmanager-main-1.alertmanager-operated:9094
    - --cluster.peer=alertmanager-main-2.alertmanager-operated:9094
    - --cluster.reconnect-timeout=5m
    env:
    - name: POD_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.podIP
…
status:
  …
  podIP: 10.145.0.149
  podIPs:
  - ip: 10.145.0.149
  - ip: fd00:abba:96fb::7a
```

But at runtime, this env var  (used in container args) is expanded to the pod IPv6 instead of the IPv4.

```
kubectl exec pod/alertmanager-main-0 -- printenv | grep POD_IP
POD_IP=fd00:abba:96fb::7a
```

```
/bin/alertmanager --config.file=/etc/alertmanager/config/alertmanager.yaml --storage.path=/alertmanager --data.retention=120h --cluster.listen-address=[fd00:abba:96fb::7a]:9094 --web.listen-address=:9093 --web.route-prefix=/ --cluster.peer
=alertmanager-main-0.alertmanager-operated:9094 --cluster.peer=alertmanager-main-1.alertmanager-operated:9094 --cluster.peer=alertmanager-main-2.alertmanager-operated:9094 --cluster.reconnect-timeout=5m
```

#### What you expected to happen:

When using `status.podIP`, it should always return the value actually exposed as `status.podIP` and not something else.

#### Environment:
- Kubernetes version:
```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-12T14:11:29Z", GoVersion:"go1.16.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:53:14Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration: Bare metal cluster
- OS (e.g: `cat /etc/os-release`): Ubuntu 20.04
- Kernel (e.g. `uname -a`): Linux 5.8.0-59-generic 66~20.04.1-Ubuntu SMP Thu Jun 17 11:14:10 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
- Network plugin and version (if this is a network-related bug): cilium 1.10
- Others: CRI-O 1.21 + crun.

**Comment by Jean-Daniel** [comment]
Note: I have a configuration that match the ones that have IP ordering issue before #95251 was fixed. 

So it may be related.

**Comment by Jean-Daniel** [comment]
Just a guess, but I wonder if it may be caused by [this code](https://github.com/kubernetes/kubernetes/blob/5f8f5ab3268b41d45481d2d865716fa04c026aab/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L816)

```
 // Overwrite the podIPs passed in the pod status, since we just started the pod sandbox.
 podIPs = m.determinePodSandboxIPs(pod.Namespace, pod.Name, podSandboxStatus)
```

**Comment by Jean-Daniel** [comment]
That's it. The fix from  #97979 properly sort IP when computing the Pod status, but the sorted IPs are then replaced by computed IPs as this log shows.

```
Jun 29 09:21:23 worker-1.kube kubelet[2312103]: I0629 09:21:23.174171 2312103 kuberuntime_manager.go:798] "Created PodSandbox for pod" podSandboxID="6ce471e2ad25ec4c29dae01bf9bc9cf156aba31bb3d8a402fdd3b7c0f56033d4" pod="default/testip"
Jun 29 09:21:23 worker-1.kube kubelet[2312103]: I0629 09:21:23.174706 2312103 kuberuntime_manager.go:817] "Determined the ip for pod after sandbox changed" IPs=[fd00:abba:96fb::14d 10.145.1.148] pod="default/testip"
```

**Comment by aojea** [comment]
/triage accepted
Thanks for the detailed report
